### PR TITLE
some more clippy-based improvements

### PR DESCRIPTION
**Pull request by llogiq**
It's a large number of small improvements to the code, mostly readability-related, but removing closures and replacing `str::to_string()` with `.to_owned()` should also positively affect performance.

r? @Manishearth (once it compiles, of course :smile:)

### CODE REVIEW

File: src/libserialize/json.rs [hunk 1]
```diff
@@ -1228,8 +1228,8 @@ impl Index<usize> for Json {
     type Output = Json;
 
     fn index<'a>(&'a self, idx: usize) -> &'a Json {
```
Comment (Manishearth) [inline src/libserialize/json.rs]: No elision?

File: src/libserialize/json.rs [hunk 2]
```diff
@@ -1323,20 +1323,20 @@ impl Stack {
...
     }
 
     /// Returns true if the bottom-most elements of this stack are the same as
     /// the ones passed as parameter.
     pub fn starts_with(&self, rhs: &[StackElement]) -> bool {
         if self.stack.len() < rhs.len() { return false; }
-        for i in 0..rhs.len() {
-            if self.get(i) != rhs[i] { return false; }
+        for (i, r) in rhs.iter().enumerate() {
+            if &self.get(i) != r { return false; }
```
Comment (Manishearth) [inline src/libserialize/json.rs]: `self.get(i) != *r`

actually, it's possible that it will coerce away

File: src/libserialize/json.rs [hunk 3]
```diff
@@ -1344,15 +1344,15 @@ impl Stack {
     pub fn ends_with(&self, rhs: &[StackElement]) -> bool {
         if self.stack.len() < rhs.len() { return false; }
         let offset = self.stack.len() - rhs.len();
-        for i in 0..rhs.len() {
-            if self.get(i + offset) != rhs[i] { return false; }
+        for (i, r) in rhs.iter().enumerate() {
+            if &self.get(i + offset) != r { return false; }
```
Comment (Manishearth) [inline src/libserialize/json.rs]: same

File: src/libserialize/json.rs [hunk 4]
```diff
@@ -1458,13 +1458,13 @@ impl<T: Iterator<Item=char>> Parser<T> {
             state: ParseStart,
         };
         p.bump();
-        return p;
+        p
     }
 
     /// Provides access to the current position in the logical structure of the
     /// JSON stream.
     pub fn stack<'l>(&'l self) -> &'l Stack {
```
Comment (Manishearth) [inline src/libserialize/json.rs]: elision?

File: src/libstd/sys/unix/os.rs [hunk 5]
```diff
@@ -142,7 +142,7 @@ pub fn split_paths<'a>(unparsed: &'a OsStr) -> SplitPaths<'a> {
     let unparsed = unparsed.as_bytes();
     SplitPaths {
         iter: unparsed.split(is_colon as fn(&u8) -> bool)
```
Comment (Manishearth) [inline src/libstd/sys/unix/os.rs]: do we even need the `as` here? Pretty sure they should just work directly a la needless_closure

File: src/libtest/stats.rs [hunk 6]
```diff
@@ -168,7 +168,7 @@ impl Stats for [f64] {
             let mut j = 0;
             // This inner loop applies `hi`/`lo` summation to each
             // partial so that the list of partial sums remains exact.
-            for i in 0..partials.len() {
+            for mut y in &mut partials {
                 let mut y: f64 = partials[i];
```
Comment (Manishearth) [inline src/libtest/stats.rs]: remove this line, won't compile otherwise

Comment (llogiq) [inline src/libtest/stats.rs]: Ouch – it's probably too late for me now; else I wouldn't have missed that one. Time to get some sleep.

**Comment by llogiq** [comment]
I didn't do all the elision things yet. We can wrap them up in the next PR.

File: src/libstd/sys/unix/os.rs (hunk 5 above)
Comment (llogiq) [inline src/libstd/sys/unix/os.rs]: Judging from the compiler error, we do need it.

**Comment by llogiq** [comment]
@Manishearth r? I may add more stuff in later PRs.

**Comment by llogiq** [comment]
Yay! :smile:

**Pull request merged**